Formalize OpenTelemetry Mission+Vision+Values (#853)
open-telemetry/community: mission-vision-values.md

# OTel Mission, Vision, and Values

## Mission (our overall north star as a community)

OpenTelemetry's Mission: **to enable effective observability through
high-quality, turnkey, open-source telemetry.**

## Vision (the world we imagine for OTel end-users)

Excellent observability requires high-quality telemetry, as well as the
performant, consistent instrumentation that makes it possible.

Telemetry in this context is the firehose of raw observational data streaming
out of software applications, and while “high-quality telemetry” may be a
requirement for excellent observability, it’s still unreasonable and
unrealistic to expect developers of application software to add or maintain the
necessary instrumentation on their own. That is a massive undertaking, and in
practice it’s not “just code” – there is also necessary alignment around
protocols and semantic conventions for tags, attributes, and other metadata to
consider.

So how do we get high-quality, turnkey telemetry without a massive,
unsustainable engineering effort? This is where OpenTelemetry comes in. To
achieve our vision, OpenTelemetry sees five key opportunities, listed here:

> **Opportunity I: Telemetry should be easy.** With OpenTelemetry we want

Reviewer: I would drop the word Opportunity from these bullets.

Author: Fair point – I think this is what we did for the "Values" below, too.

high-quality telemetry to be easy, especially for end-users. That means that
OpenTelemetry should have fast time-to-value, set reasonable defaults yet allow
for customization, and pair with excellent documentation and a top-tier overall
developer experience.

**Opportunity II: Telemetry should be universal.** Telemetry protocols and
conventions should be unified across languages and signal types (tracing,
metrics, logging, etc), not divergent or siloed. This means that OpenTelemetry
aspires to find technical solutions that work consistently, both locally and
globally.

> **Opportunity III: Telemetry should be vendor-neutral.** For decades, monitoring

Reviewer 2: In light of recent security breaches related to agents, should we also highlight that OTel's vendor-neutral OSS approach is more auditable?

providers, avoiding lock-in for any vendor and interoperating with other OSS
projects in the telemetry and observability ecosystem.

**Opportunity IV: Telemetry should be loosely coupled.** OpenTelemetry end-users
should be able to pick and choose from the pieces they want without bringing in
the rest of the project. To enable this, OpenTelemetry’s software architecture
is decoupled wherever possible. As a corollary, this also means that
OpenTelemetry does not want to “pick winners” when it comes to particular
projects or technologies: where possible, we prefer to create optionality for
our end-users.

**Opportunity V: Telemetry should be built-in.** Historically, telemetry was
something developers integrated manually or via post-compilation agents.
OpenTelemetry believes that high-quality telemetry can be built in to the
entire software stack – just like testing has been for decades.

While the structure and technical details of OpenTelemetry may change over
time, these five key opportunities will remain outstanding until we achieve our
mission, and as a project we refer to them to orient – and reorient – as we
chart our path.

> ## Engineering Values (the principles that guide our contributions)

Reviewer 2: These values describe the artifacts we produce. Do we also want to document values we hold as an engineering community/organization? E.g.,
* bias towards action and experimentation vs consensus-building
* inclusion, mentorship and investment in new contributors
* respect for maintainers

Reviewer 3: these values were covered in https://docs.google.com/document/d/1L92ThlgvSxYlQixpBNEovz8NnyOZfKu1E3zTT2gxaCY/edit#heading=h.id952e4ms1v7 under execution principles. Including those may be too much for this document. Do you feel this have to be formulated as values?

Author: Yeah, my concern was brevity. I think it's a great idea to document "the way we work", but my thought process was that "the way we work" might change significantly over time as the project matures, and as such it doesn't make it into this doc which is meant to be for more "permanent" / unchanging values. Thoughts?

Reviewer 2: If that material will go to a different, prominent location then I'm 👍🏽

OpenTelemetry’s mission and vision describe where we want to go.
OpenTelemetry’s engineering values describe how we want to get there.

> OpenTelemetry’s core engineering values are Compatibility, Stability,

Reviewer 4: I wish a user focus ("ease of use"? "usability"?) were a part of these engineering values

Author: "Telemetry should be easy" is the first part of the vision... is that sufficient?

Reviewer 4: I'd say probably not - I think vision generally drives high level strategy / decision making while values are more for day-to-day. I hope engineers on the project value user experience in all their work and don't get that sense from the current wording. Of course this is just my hope - it's perfectly possible for my own values to not align with the project, it's up to the project to decide what to put here :P

Resiliency, and Performance.

### We Value *Compatibility*

> With the number of stakeholders and supported platforms, following

Reviewer: > With the number of stakeholders and supported platforms,

incomplete clause? "with the number of stakeholders" doing what? Maybe "given the number..."

> specifications and enabling interoperability is very important. This means that

Reviewer: > This means that

unnecessary filler

OpenTelemetry strives to be standards-compliant, vendor-neutral, and consistent
across languages and components.

### We Value *Stability*

As many libraries take dependencies on OpenTelemetry APIs, API stability and
backwards compatibility is vital for our end-users, and thus for our work. As a

> corollary, we do not introduce new concepts unless we’re confident they’re

Reviewer: > and thus for our work. As a corollary

unnecessary filler

needed by a broad subset of OpenTelemetry’s end-users.

### We Value *Resilience*

> In OpenTelemetry we value technical resiliency: the ability to adapt to

Reviewer: ```suggestion
We value technical resiliency: the ability to adapt and to
```

continue operating even in the face of resource scarcity or other environmental
challenges. OpenTelemetry is designed to work and keep collecting telemetry
signals when an application is misbehaving, and OpenTelemetry code is designed
to degrade gracefully as needed.

### We Value *Performance*

> The cardinal rule in observability is “first, do no harm”: to that end,

Reviewer: "do no harm" to me also means "don't crash my app", which is not "performance", and not exactly the same as "resilience" above

Reviewer 5: Perhaps we could merge these sections

Reviewer 2: Related: in the face of constraints, business logic should always take precedence over telemetry collection.

Reviewer 3: let's not merge these sections. Performance is intentionally value number 4, less important than 3 ahead of it. Maybe just rephrase that it is more about ROI of telemetry than "no harm".

# OTel Mission, Vision, and Values

> ## Mission (our overall north star as a community)

Reviewer 6: Is the order of vision and mission what we want? I'm used to seeing the mission following a vision since a mission's purpose is to help realize a vision.

Author: I guess I'm accustomed (`N=1` here, not saying I'm correct!) that the "Mission" goes first, and is typically very short and easy to memorize/reference. The vision is "below the fold" for those who want to dig deeper.

### Telemetry should be easy

With OpenTelemetry we want high-quality telemetry to be easy, especially for
end-users. That means that OpenTelemetry should have fast time-to-value, set
reasonable defaults yet allow for customization, and pair with excellent
documentation and a top-tier overall developer experience.

### Telemetry should be universal

Telemetry protocols and conventions should be unified across languages and
signal types (tracing, metrics, logging, etc), not divergent or siloed. This
means that OpenTelemetry aspires to find technical solutions that work
consistently, both locally and globally.

### Telemetry should be vendor-neutral

For decades, monitoring and observability vendors have created differentiation
– and lock-in – through agents that automatically gather useful telemetry from
a wide integration surface. This lock-in is limiting for customers and inhibits
innovation by tightly coupling telemetry collection and telemetry storage and
analysis. With OpenTelemetry, we strive to provide a level playing field for
all observability providers, avoiding lock-in for any vendor and interoperating
with other OSS projects in the telemetry and observability ecosystem.

> Furthermore, with the looming threat of software-supply-chain attacks, we aim

Reviewer 6: This last line referencing software-supply-chain-attacks feels like a periphery line item to a larger subject of vendor-neutral.  I've not seen a statement about software security. Honestly, I may have missed it as I read through this doc.

Author: yeah, fair point – this sentence was added in response to an earlier reviewer comment. I'll try to allude to the broader benefits of OSS without getting so far into the weeds on this.

# OTel Mission, Vision, and Values

## Mission (our overall north star as a community)

> OpenTelemetry's Mission: **to enable effective observability through

Author: Other options from the gdoc – I welcome suggestions here!

```
to make high-quality telemetry a built-in feature for all cloud-native software.
to enable high-quality observability through ubiquitous, portable, high-quality telemetry
to enable effective observability through built-in, portable, open-source telemetry
to enable effective observability by making high-quality telemetry a built-in feature for all cloud-native software
to enable effective observability through turnkey telemetry
to enable effective observability through high-quality, turnkey telemetry
to enable observability through high-quality, turnkey, and ubiquitous telemetry
to enable observability through high-quality, open source telemetry
to enable observability through high-quality telemetry
to enable excellent observability through high-quality telemetry
```

Reviewer 7: I like "to make high-quality telemetry a built-in feature for all cloud-native software." 

To me the defining feature of our mission is to get frameworks to build in first party support. It's what makes us most different from other solutions.

Reviewer 8: what @dyladan said, we need "cloud-native" to be our primary remit (but also we shouldn't exclude other users).

wrt including "observability" or not, I think the telemetry is a crucial ingredient, but adopting otel alone won't give you o11y

Author: I think I wrote `to make high-quality telemetry a built-in feature for all cloud-native software` and so can still get behind it personally. :)

The critique – from @yurishkuro IIRC – was:

>This to me reads more like a Vision as defined in the About section, an "ideal state", and as such not appropriate as Mission because then mission just says "implement Vision"

In order to avoid the tautological aspect, I was trying to do a sort of "five whys" exercise and ask myself what high-quality telemetry is *good* for, hence **enabling** observability (rather than **being** o11y, @lizthegrey). Maybe that distinction is too subtle, though.

If we don't want to say "observability" here, are we able to zoom out and articulate a larger purpose for OpenTelemetry beyond, well, telemetry?

Separately regarding "cloud-native": as a litmus test, would we welcome the news that old-school networking hardware is going to support OTLP export of load metrics? I would... but that's not necessarily "cloud-native." Of course OTel does *include* cloud-native, but do we want to be *exclusively* cloud-native? And if not, why is it in the mission?

Reviewer: I think the current text is confusing Vision (ideal state) and Mission (how to get there). Vision come first.

Reviewer 9: My vote is for: "to enable high-quality observability through ubiquitous, portable, high-quality telemetry"  for the reason mentioned by @bhs above.

 I would like to see OTel be adopted ubiquitously with no implication that its a "cloud thing". For me explicitly adding "cloud-native" is problematic.  I see a challenge that initiatives like OTel can meet is providing that ubiquity between cloud and non-cloud use cases. Ubiquity is the key principle for me. 

It may be semantics but As some one who worries about both worlds, being able to reason about observability as a singular thing across multiple domains is a big strength that I would encourage highlighting rather than ring fencing with the term "cloud-native"

high-quality, turnkey, open-source telemetry.**

## Vision (the world we imagine for OTel end-users)

Excellent observability requires high-quality telemetry, as well as the
performant, consistent instrumentation that makes it possible.

Telemetry in this context is the firehose of raw observational data streaming
out of software applications, and while “high-quality telemetry” may be a
requirement for excellent observability, it’s still unreasonable and

> unrealistic to expect software developers to add or maintain the necessary

Reviewer 3: software developers in this case are not library owners. Maybe this needs to be highlighted somehow.

Reviewer 3: by library I mean client libraries that needs to be instrumented with OTel

Reviewer 3: need to distinguish these software developers and end users

Author: I made this "developers of application software" – good call-out.

Reviewer 10: I like "developers of application software".

semantic conventions for tags, attributes, and other metadata to consider.

So how do we get high-quality, turnkey telemetry without a massive,
unsustainable engineering effort? This is where OpenTelemetry comes in. To
achieve our vision, OpenTelemetry sees five key opportunities, listed here:

**Opportunity I: Telemetry should be easy.** With OpenTelemetry we want
high-quality telemetry to be easy, especially for end-users. That means that
OpenTelemetry should have fast time-to-value, set reasonable defaults yet allow
for customization, and pair with excellent documentation and a top-tier overall
developer experience.

**Opportunity II: Telemetry should be universal.** Telemetry protocols and
conventions should be unified across languages and signal types (tracing,
metrics, logging, etc), not divergent or siloed. This means that OpenTelemetry
aspires to find technical solutions that work consistently, both locally and
globally.

**Opportunity III: Telemetry should be vendor-neutral.** For decades, monitoring
and observability vendors have created differentiation – and lock-in – through
agents that automatically gather useful telemetry from a wide integration

> surface. While convenient, this lock-in is limiting for customers and inhibits

Reviewer 11: >While convenient

I think this sentence works better without these 2 words.

Reviewer 7: Agree with tigran here. no need to soften this sentence

Author: great point, removed!

innovation by tightly coupling telemetry collection and telemetry storage and
analysis. With OpenTelemetry, we strive to provide a level playing field for
all observability providers, avoiding lock-in for any vendor and interoperating
with other OSS projects in the telemetry and observability ecosystem.

**Opportunity IV: Telemetry should be loosely coupled.** OpenTelemetry end-users
should be able to pick and choose from the pieces they want without bringing in
the rest of the project. To enable this, OpenTelemetry’s software architecture
is decoupled wherever possible. As a corollary, this also means that
OpenTelemetry does not want to “pick winners” when it comes to particular
projects or technologies: where possible, we prefer to create optionality for
our end-users.

**Opportunity V: Telemetry should be built-in.** Historically, telemetry was
something developers integrated manually or via post-compilation agents.
OpenTelemetry believes that high-quality telemetry can be built in to the

> entire software stack – just like testing has been for decades.

Reviewer 11: I find the comparison with testing a bit confusing, not sure if it helps in any way (and it surely distract me since as I read this I try to find the similarities and differences between testing and observability :-) ). Perhaps drop the reference to testing.

Reviewer 8: disagree - it's an analogy to writing comments, or writing unit tests - that it should be easy to do so, so that it shifts left.

Author: (I eagerly wait additional opinions on this matter :)

Reviewer: I see what Liz is saying, but the text does not read like that

Reviewer 8: the phrasing _I_ like to use that we can steal is that writing code without comments or tests would be unthinkable, so code without telemetry should similarly be unthinkable.

Reviewer: > writing code without comments or tests would be unthinkable

I wish this were true, but I am onboard with the sentiment. After all, we're describing the vision.

Reviewer 5: +1 to the reference to testing being somewhat confusing (_testing_ isn't necessarily built-in to JDBC or Express.js). But I do strongly agree with the tenet of observability via OpenTelemetry being built-in to software packages.

Reviewer 2: Is CI a better analogy here?

25 years ago, the expectation was that a library consisted of a tarball and release notes. Today with Github and its ecosystem, we expect that tests run regularly, can be reviewed by contributors and adopters, and are subject to flakiness hygiene.

While the structure and technical details of OpenTelemetry may change over
time, these five key opportunities will remain outstanding until we achieve our
mission, and as a project we refer to them to orient – and reorient – as we
chart our path.

## Engineering Values (the principles that guide our contributions)

OpenTelemetry’s mission and vision describe where we want to go.
OpenTelemetry’s engineering values describe how we want to get there.

OpenTelemetry’s core engineering values are Compatibility, Stability,
Resiliency, and Performance.

### We Value *Compatibility*

With the number of stakeholders and supported platforms, following
specifications and enabling interoperability is very important. This means that
OpenTelemetry strives to be standards-compliant, vendor-neutral, and consistent
across languages and components.

### We Value *Stability*

As many libraries take dependencies on OpenTelemetry APIs, API stability and
backwards compatibility is vital for our end-users, and thus for our work. As a
corollary, we do not introduce new concepts unless we’re confident they’re
needed by a broad subset of OpenTelemetry’s end-users.

> ### We Value *Resiliency*

Reviewer 8: use "resilience" not "resiliency"
